### ISSUE TITLE
getUserData depends on getSession, usage is invalid

### ISSUE THREAD

**Issue by andrew-aladev** [aws-amplify/amplify-js#1353]
Hello. I couldn't understand a logic of `currentUserPoolUser` from [packages/auth/src/Auth.ts#L765](https://github.com/aws-amplify/amplify-js/blob/master/packages/auth/src/Auth.ts#L765). Let's try to describe it.

```
// refresh the session if the session expired.
user.getSession(...
...
// get user data from Cognito, also to make sure the user is still valid
user.getUserData(...
....
```

These functions are [here](https://github.com/aws-amplify/amplify-js/blob/master/packages/amazon-cognito-identity-js/src/CognitoUser.js#L1118):

```
getSession(callback) {
if (this.signInUserSession != null && this.signInUserSession.isValid()) {
  return callback(null, this.signInUserSession);
}
...
this.refreshSession(refreshToken, callback);
...
```

```
getUserData(callback) {
if (!(this.signInUserSession != null && this.signInUserSession.isValid())) {
  return callback(new Error('User is not authenticated'), null);
}
... getting user ...
```

When `this.signInUserSession != null && this.signInUserSession.isValid()` is `false`:
1. `getSession` is refreshing session.
2. `getUserData` throws error.
3. I am sitting here and waiting when cognito will send me a reasonable error about session.
4. I can see that cognito sent me an error: `{"__type":"PasswordResetRequiredException","message":"Password reset required for the user"}`, but promise is already rejected with `User is not authenticated`.

There is an obvious fix: call `getUserData` after `await getSession`.

But I am not an amazon developer. I shouldn't walk so far away from public api. Both `auth`, `amazon-cognito-identity-js` are **core amplify-js packages** and they are very broken.

Customer won't let me waste so much time on level such mountains. Is there any **alternative for amplify-js**?

**Comment by andrew-aladev** [comment]
We can see `checkUser` function usage for example in [packages/aws-amplify-react/src/Auth/Greetings.jsx](https://github.com/aws-amplify/amplify-js/blob/master/packages/aws-amplify-react/src/Auth/Greetings.jsx#L100). 

```
Auth.currentAuthenticatedUser()
  .catch(err => {
    this.signOut();
  });
```

This solution is completely wrong. You are forcing user to sign out + sign in without any reason. This usage should be something like this:

```
Auth.currentAuthenticatedUser()
  .catch(err => {
    switch (err.code) {
      case 'PasswordResetRequiredException':
        ...
      default:
        this.signOut();
    }
  });
```

This code will force you to reproduce the same issue that I've tried to describe here.

**Comment by andrew-aladev** [comment]
[Commit](https://github.com/aws-amplify/amplify-js/commit/1b09e2f976ad278b3d037e01624bef8a1ec3c310) @powerful23 

Please mock async functions with something like `setImmediate`:

```
const spyon2 = jest.spyOn(CognitoUser.prototype, 'getSession')
  .mockImplementation((callback) => {
    setImmediate(() => callback(null, 'session'));
});
```

Please include unhappy scenarios like:

```
const spyon3 = jest.spyOn(CognitoUser.prototype, 'getUserData')
  .mockImplementation((callback) => {
    setImmediate(() => callback(new Error('User is not authenticated'), null));
});
```

It looks like you are implementing all these tests from scratch. I want to say that unit tests will never give you a good results, auth system needs integration tests.

Let's try to ask official project manager. @kaustavghosh06, hello. Cognito auth system has a big amount of SDK, almost for any programming language. Is there any formalized requirements spec that could be used to implement integration tests? Do you ever have any plan to implement something more serious than unit tests. Thank you.

**Comment by andrew-aladev** [comment]
Today we've used `"amazon-cognito-identity-js": "^1.31.0"` and implemented very small and simplified version of `Auth.ts`. Authentication flow works good.

**Comment by volodymyrrudyi** [comment]
Same here, basically AWS Amplify doesn't handle expired tokens correctly.

**Comment by tcchau** [comment]
I have not even gotten to the point where I'm handling scenarios like password reset required; have had to spend an afternoon stepping into Amplify JS code to arrive at the same conclusion that there's essentially a race condition set up by the ```currentUserPoolUser``` call: the call to ```getUserData``` will always fail at least once. 

This make it impossible to refresh the user's session when they come back to our website/app because initially it would be as if they are an unauthenticated user. The user has to then reload the website/app for it to work again, because at this point, the ```InitiateAuth``` call to AWS Cognito will have completed, and the new access, ID and refresh tokens have been retrieved, and stored in local storage.

Now... how do I find a workaround?

**Comment by powerful23** [comment]
@tcchau can you provide some detailed info about your info like how you are using the library and what's the error when calling `currentUserPoolUser`. Also recently we did some fix around that method so can you ensure you are using the latest version?

**Comment by tcchau** [comment]
@powerful23 In fact, I checked the code in Auth.ts in the master branch and it does appear that it is fixed. Essentially, ```getUserData``` is called in the callback passed into ```getSession```. Thus, the library will only attempt to ```getUserData``` after ```getSession``` is completed, rather than immediately after the ```getSession``` call as it was previously. 

From the blame, it actually looks like you fixed this 14 days ago as a part of a different fix (https://github.com/aws-amplify/amplify-js/commit/c61505a9e935aa5cb5d7131f106787c47c03bb92).

However, version 1.0.8 in the NPM registry doesn't incorporate this fix yet, it seems.

Would you know when another release to NPM will take place?

**Comment by powerful23** [comment]
@tcchau the fixed version should be published as `aws-amplify@1.0.10`

**Comment by tcchau** [comment]
@powerful23 Cool, I see it now! I will give it a try and report back if there are problems.
